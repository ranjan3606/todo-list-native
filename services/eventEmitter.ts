type EventCallback = (...args: any[]) => void;

class EventEmitter {
  private events: Record<string, EventCallback[]> = {};

  public on(event: string, callback: EventCallback): () => void {
    if (!this.events[event]) {
      this.events[event] = [];
    }

    this.events[event].push(callback);

    // Return a function to unsubscribe
    return () => {
      this.events[event] = this.events[event].filter(cb => cb !== callback);
    };
  }

  public emit(event: string, ...args: any[]): void {
    if (!this.events[event]) {
      return;
    }

    this.events[event].forEach(callback => {
      callback(...args);
    });
  }
}

// Create a singleton instance
export const todoEvents = new EventEmitter();

// Define event types
export const TODO_EVENTS = {
  TODO_ADDED: 'todo_added',
  TODO_UPDATED: 'todo_updated',
  TODO_DELETED: 'todo_deleted',
  TODO_COMPLETED: 'todo_completed',
  TODO_SNOOZED: 'todo_snoozed',
  STORAGE_CHANGED: 'storage_changed',
  // Add new tag event types
  TAG_ADDED: 'tag_added',
  TAG_UPDATED: 'tag_updated',
  TAG_DELETED: 'tag_deleted',
  TAGS_CHANGED: 'tags_changed'
};
