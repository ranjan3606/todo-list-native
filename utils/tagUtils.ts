/**
 * Utility functions for tag management
 */

/**
 * Format a list of keywords to ensure they start with #
 * @param keywords Array of keywords to format
 * @returns Array of formatted keywords
 */
export const formatTagKeywords = (keywords: string[]): string[] => {
  return keywords
    .map(k => k.trim())
    .filter(k => k.length > 0)
    .map(k => k.startsWith('#') ? k : `#${k}`);
};

/**
 * Convert comma-separated string to array of keywords
 * @param keywordsString Comma-separated keywords string
 * @returns Array of formatted keywords
 */
export const parseKeywordsString = (keywordsString: string): string[] => {
  const keywords = keywordsString
    .split(',')
    .map(k => k.trim())
    .filter(k => k.length > 0);
    
  return formatTagKeywords(keywords);
};

/**
 * Create animation values for a set of tags
 * @param tags Object containing tag names as keys
 * @param currentAnimValues Existing animation values
 * @param defaultValue Default animation value
 * @returns Updated animation values object
 */
export const initializeTagAnimations = (
  tags: Record<string, any>,
  currentAnimValues: Record<string, any>,
  defaultValue: number
): Record<string, any> => {
  const updatedAnimValues = { ...currentAnimValues };
  
  Object.keys(tags).forEach(tagName => {
    if (!updatedAnimValues[tagName]) {
      updatedAnimValues[tagName] = new Animated.Value(defaultValue);
    }
  });
  
  return updatedAnimValues;
};
