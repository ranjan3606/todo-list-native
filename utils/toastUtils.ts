import { Platform } from 'react-native';
import * as Haptics from 'expo-haptics';
import { todoEvents } from '@/services/eventEmitter';

// Event name for our toast
const SHOW_TOAST_EVENT = 'SHOW_TOAST';

// Toast types
export type ToastType = 'success' | 'error' | 'info';

// Toast message structure
export interface ToastMessage {
  message: string;
  type?: ToastType;
  duration?: number;
}

/**
 * Shows a toast notification
 * @param message Message to display or a toast object with additional configuration
 * @param type Optional toast type (success, error, info)
 * @param duration Optional duration in milliseconds
 */
export const showToast = (
  message: string | ToastMessage,
  type: ToastType = 'success',
  duration: number = 3000
) => {
  // Trigger haptic feedback
  if (Platform.OS !== 'web') {
    Haptics.notificationAsync(
      type === 'error' 
        ? Haptics.NotificationFeedbackType.Error
        : Haptics.NotificationFeedbackType.Success
    );
  }

  // Format the message object
  const toastMessage: ToastMessage = typeof message === 'string'
    ? { message, type, duration }
    : { ...message };

  // Emit the event with the toast data
  todoEvents.emit(SHOW_TOAST_EVENT, toastMessage);
};

// Export the event name for listeners
export { SHOW_TOAST_EVENT };