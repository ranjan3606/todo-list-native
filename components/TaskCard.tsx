import React, { useState } from 'react';
import { View, Text, TouchableOpacity, StyleSheet, Alert, Modal, SafeAreaView } from 'react-native';
import { FontAwesome } from '@expo/vector-icons';
import { Colors } from '@/constants/Colors';
import { Todo } from '@/types/todo';
import { SwipeableRow } from './SwipeableRow';
import { formatDisplayDate, isToday, isTomorrow, isDueSoon, getNextDueDate } from '@/utils/dateUtils';
import { deleteTodo, toggleTodoCompletion, snoozeTodo, getSnoozeDuration, addTodo } from '@/services/storage';
import { useTranslation } from '@/i18n';

interface TaskCardProps {
  item: Todo;
  colorScheme: 'light' | 'dark';
  onEdit?: (task: Todo) => void;
}

export const TaskCard: React.FC<TaskCardProps> = ({ 
  item, 
  colorScheme,
  onEdit
}) => {
  const { t } = useTranslation();
  const [isProcessing, setIsProcessing] = useState(false);
  const [showDetails, setShowDetails] = useState(false);
  const isDueToday = isToday(item.dueDate);
  const isDueTomorrow = isTomorrow(item.dueDate);
  const isDue = isDueSoon(item.dueDate);

  const handleLongPress = () => {
    if (onEdit) {
      Alert.alert(
        t('task.edit'),
        t('task.deleteConfirm'),
        [
          { text: t('common.cancel'), style: "cancel" },
          { text: t('common.edit'), onPress: () => onEdit(item) },
          { text: t('common.delete'), onPress: () => handleDelete(), style: "destructive" }
        ]
      );
    }
  };

  const handleDelete = async () => {
    if (isProcessing) return;
    
    setIsProcessing(true);
    try {
      const success = await deleteTodo(item.id);
      if (!success) {
        Alert.alert(t('common.error'), t('taskForm.error'));
      }
      setShowDetails(false);
    } catch (error) {
      Alert.alert(t('common.error'), t('taskForm.error'));
    } finally {
      setIsProcessing(false);
    }
  };

  const handleComplete = async () => {
    if (isProcessing) return;
    
    setIsProcessing(true);
    try {
      const result = await toggleTodoCompletion(item.id);
      
      // Handle recurring task completion
      if (result.success && result.newTodo && result.newTodo.completed && 
          result.newTodo.recurring && result.newTodo.recurring !== 'none') {
        
        const nextDate = getNextDueDate(result.newTodo.dueDate, result.newTodo.recurring);
        
        if (nextDate !== result.newTodo.dueDate) {
          const nextTodoInstance: Todo = {
            ...result.newTodo,
            id: Date.now().toString(),
            dueDate: nextDate,
            completed: false,
            originalDueDate: result.newTodo.originalDueDate || result.newTodo.dueDate,
            isRecurringInstance: true,
          };
          
          // Add the next instance to storage
          await addTodo(nextTodoInstance);
        }
      }
      
      setShowDetails(false);
    } catch (error) {
      Alert.alert(t('common.error'), t('task.error'));
    } finally {
      setIsProcessing(false);
    }
  };

  const handleSnooze = async () => {
    if (isProcessing) return;
    
    setIsProcessing(true);
    try {
      const snoozeDuration = await getSnoozeDuration();
      const result = await snoozeTodo(item.id, snoozeDuration);
      
      if (result.success && result.newDueDate) {
        const hours = snoozeDuration === 1 ? 
          t('task.hour') : 
          `${snoozeDuration} ${t('task.hours')}`;
        
        Alert.alert(
          t('task.snoozed'), 
          t('task.snoozedFor', { hours })
        );
      } else {
        Alert.alert(t('common.error'), t('taskForm.error'));
      }
      
      setShowDetails(false);
    } catch (error) {
      Alert.alert(t('common.error'), t('taskForm.error'));
    } finally {
      setIsProcessing(false);
    }
  };

  // Task detail popup component
  const TaskDetailPopup = () => (
    <Modal
      visible={showDetails}
      transparent={true}
      animationType="fade"
      onRequestClose={() => setShowDetails(false)}
    >
      <TouchableOpacity 
        style={styles.modalOverlay} 
        activeOpacity={1} 
        onPress={() => setShowDetails(false)}
      >
        <View 
          style={[
            styles.modalContent, 
            { backgroundColor: Colors[colorScheme].cardBackground }
          ]}
          onStartShouldSetResponder={() => true}
          onTouchEnd={(e) => e.stopPropagation()}
        >
          {/* Task title and description */}
          <Text style={[styles.modalTitle, { color: Colors[colorScheme].text }]}>
            {item.name}
          </Text>
          
          {item.description ? (
            <Text style={[styles.modalDescription, { color: Colors[colorScheme].text }]}>
              {item.description}
            </Text>
          ) : null}
          
          {/* Due date info */}
          {item.dueDate && (
            <View style={styles.modalInfoRow}>
              <FontAwesome name="calendar" size={16} color={Colors[colorScheme].lightText} />
              <Text style={[styles.modalInfoText, { color: Colors[colorScheme].text }]}>
                {t('task.due', { date: formatDisplayDate(item.dueDate) })}
              </Text>
            </View>
          )}
          
          {/* Recurring info */}
          {item.recurring && item.recurring !== 'none' && (
            <View style={styles.modalInfoRow}>
              <FontAwesome name="repeat" size={16} color={Colors[colorScheme].lightText} />
              <Text style={[styles.modalInfoText, { color: Colors[colorScheme].text }]}>
                {t(`taskForm.${item.recurring}`)}
              </Text>
            </View>
          )}
          
          {/* Original due date for recurring tasks */}
          {item.isRecurringInstance && item.originalDueDate && (
            <View style={styles.modalInfoRow}>
              <FontAwesome name="history" size={16} color={Colors[colorScheme].lightText} />
              <Text style={[styles.modalInfoText, { color: Colors[colorScheme].text }]}>
                {t('task.originally', { date: formatDisplayDate(item.originalDueDate) })}
              </Text>
            </View>
          )}
          
          {/* Tags */}
          {item.tags && item.tags.length > 0 && (
            <View style={styles.modalTagsContainer}>
              {item.tags.map(tag => (
                <Text key={tag} style={[
                  styles.itemTag,
                  { 
                    color: Colors[colorScheme].tint,
                    backgroundColor: `${Colors[colorScheme].tint}15`,
                  }
                ]}>
                  {tag}
                </Text>
              ))}
            </View>
          )}
          
          {/* Action buttons */}
          <View style={styles.modalActions}>
            <TouchableOpacity 
              style={[styles.modalButton, { backgroundColor: Colors[colorScheme].tint }]}
              onPress={() => {
                setShowDetails(false);
                if (onEdit) onEdit(item);
              }}
            >
              <FontAwesome name="edit" size={18} color="#fff" />
              <Text style={styles.modalButtonText}>{t('common.edit')}</Text>
            </TouchableOpacity>
            
            <TouchableOpacity 
              style={[styles.modalButton, { backgroundColor: item.completed ? Colors.primary : Colors.success }]}
              onPress={handleComplete}
              disabled={isProcessing}
            >
              <FontAwesome name={item.completed ? "undo" : "check"} size={18} color="#fff" />
              <Text style={styles.modalButtonText}>
                {item.completed ? t('task.unmark') : t('task.markComplete')}
              </Text>
            </TouchableOpacity>
            
            <TouchableOpacity 
              style={[styles.modalButton, { backgroundColor: Colors.warning }]}
              onPress={handleSnooze}
              disabled={isProcessing}
            >
              <FontAwesome name="clock-o" size={18} color="#fff" />
              <Text style={styles.modalButtonText}>{t('task.snooze')}</Text>
            </TouchableOpacity>
            
            <TouchableOpacity 
              style={[styles.modalButton, { backgroundColor: Colors.danger }]}
              onPress={handleDelete}
              disabled={isProcessing}
            >
              <FontAwesome name="trash" size={18} color="#fff" />
              <Text style={styles.modalButtonText}>{t('common.delete')}</Text>
            </TouchableOpacity>
          </View>
        </View>
      </TouchableOpacity>
    </Modal>
  );

  const renderCardContent = () => (
    <TouchableOpacity 
      style={[
        styles.todoItem, 
        { backgroundColor: Colors[colorScheme].cardBackground },
        isDueToday && styles.todayTask,
        isDueTomorrow && styles.tomorrowTask,
        isDue && !isDueToday && !isDueTomorrow && styles.upcomingTask,
        item.completed && styles.completedTask
      ]}
      onLongPress={handleLongPress}
      delayLongPress={500}
      onPress={() => setShowDetails(true)} // Changed to open detail popup instead of toggling completion
    >
      <View style={styles.todoContent}>
        {/* Completion Icon */}
        <TouchableOpacity 
          onPress={(e) => {
            e.stopPropagation(); // Prevent the card's onPress from firing
            handleComplete();
          }} 
          style={styles.checkIcon}
        >
          <FontAwesome
            name={item.completed ? "check-circle" : "circle-o"}
            size={24}
            color={item.completed ? Colors.success : (isDue ? Colors[colorScheme].dueSoon : "#ccc")}
          />
        </TouchableOpacity>
        
        {/* Task Text and Info */}
        <View style={styles.todoTextContainer}>
          <Text style={[
            styles.todoName, 
            { color: Colors[colorScheme].text },
            item.completed && styles.completedText
          ]}>
            {item.name}
          </Text>
          
          {item.description ? (
            <Text style={[
              styles.todoDesc, 
              { color: Colors[colorScheme].lightText },
              item.completed && styles.completedText
            ]}>
              {item.description}
            </Text>
          ) : null}
          
          {/* Display Due Date and Recurrence */}
          <View style={styles.dateRecurrenceContainer}>
            {item.dueDate ? (
              <Text style={[
                styles.todoDate,
                { 
                  color: isDueToday && !item.completed ? 
                    Colors[colorScheme].dueSoon : 
                    Colors[colorScheme].lightText 
                },
                item.completed && styles.completedText
              ]}>
                {t('task.due', { date: formatDisplayDate(item.dueDate) })}
              </Text>
            ) : null}
            
            {item.isRecurringInstance && item.originalDueDate && (
              <Text style={[
                styles.todoDate, 
                { 
                  fontStyle: 'italic', 
                  marginLeft: 5,
                  color: Colors[colorScheme].lightText
                },
                item.completed && styles.completedText
              ]}>
                {t('task.originally', { date: formatDisplayDate(item.originalDueDate) })}
              </Text>
            )}
            
            {item.recurring && item.recurring !== 'none' && (
              <Text style={[
                styles.recurringText, 
                { color: Colors[colorScheme].lightText },
                item.completed && styles.completedText
              ]}>
                <FontAwesome name="repeat" size={12} color={Colors[colorScheme].lightText} /> 
                {t(`taskForm.${item.recurring}`)}
              </Text>
            )}
          </View>
          
          {/* Display Tags */}
          {item.tags && item.tags.length > 0 && (
            <View style={styles.itemTagContainer}>
              {item.tags.map(tag => (
                <Text key={tag} style={[
                  styles.itemTag,
                  { 
                    color: Colors[colorScheme].tint,
                    backgroundColor: `${Colors[colorScheme].tint}15`,
                    opacity: item.completed ? 0.7 : 1
                  },
                  item.completed && styles.completedText
                ]}>
                  {tag}
                </Text>
              ))}
            </View>
          )}
        </View>
      </View>
    </TouchableOpacity>
  );

  return (
    <>
      <SwipeableRow 
        onDelete={handleDelete}
        onComplete={handleComplete}
        onSnooze={handleSnooze}
        isCompleted={item.completed}
      >
        {renderCardContent()}
      </SwipeableRow>
      
      {/* Detail Popup */}
      <TaskDetailPopup />
    </>
  );
};

const styles = StyleSheet.create({
  todoItem: {
    padding: 15,
    borderBottomWidth: 1,
    borderBottomColor: '#eee',
    borderRadius: 5,
  },
  completedTask: {
    opacity: 0.8,
  },
  todoContent: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
    marginRight: 10,
  },
  checkIcon: {
    paddingRight: 10,
  },
  todoTextContainer: {
    flex: 1,
  },
  todoName: {
    fontSize: 18,
    fontWeight: 'bold',
  },
  completedText: {
    textDecorationLine: 'line-through',
    opacity: 0.7,
  },
  todoDesc: {
    marginTop: 5,
  },
  todoDate: {
    marginTop: 5,
  },
  dateRecurrenceContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginTop: 3,
    flexWrap: 'wrap',
  },
  recurringText: {
    marginLeft: 10,
    fontSize: 12,
    fontStyle: 'italic',
  },
  itemTagContainer: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    marginTop: 5,
  },
  itemTag: {
    fontSize: 11,
    paddingHorizontal: 6,
    paddingVertical: 2,
    borderRadius: 8,
    marginRight: 5,
    marginBottom: 3,
  },
  todayTask: {
    borderLeftWidth: 4,
    borderLeftColor: Colors.danger,
  },
  tomorrowTask: {
    borderLeftWidth: 4,
    borderLeftColor: Colors.warning,
  },
  upcomingTask: {
    borderLeftWidth: 4,
    borderLeftColor: Colors.info,
  },
  // Modal styles
  modalOverlay: {
    flex: 1,
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20,
  },
  modalContent: {
    width: '90%',
    maxWidth: 400,
    borderRadius: 10,
    padding: 20,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
  },
  modalTitle: {
    fontSize: 20,
    fontWeight: 'bold',
    marginBottom: 10,
  },
  modalDescription: {
    fontSize: 16,
    marginBottom: 15,
    lineHeight: 22,
  },
  modalInfoRow: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 8,
  },
  modalInfoText: {
    marginLeft: 10,
    fontSize: 14,
  },
  modalTagsContainer: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    marginTop: 5,
    marginBottom: 15,
  },
  modalActions: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'space-between',
    marginTop: 15,
  },
  modalButton: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
    borderRadius: 5,
    marginBottom: 10,
    width: '48%',
  },
  modalButtonText: {
    color: '#fff',
    fontWeight: '600',
    marginLeft: 8,
  },
});
