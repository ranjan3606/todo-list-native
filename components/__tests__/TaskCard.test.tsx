import React from 'react';
import { render, fireEvent, act } from '@testing-library/react-native';
import { Alert, Modal, View, Text } from 'react-native';
import { TaskCard } from '../TaskCard';
import { deleteTodo, toggleTodoCompletion, snoozeTodo } from '@/services/storage';
import { getSnoozeDuration } from '@/services/storage';
import { showToast } from '@/utils/toastUtils';

// Fix the FontAwesome mock to avoid out-of-scope variable references
jest.mock('@expo/vector-icons', () => ({
  FontAwesome: jest.fn().mockImplementation((props) => ({
    type: 'FontAwesome',
    props: {...props, testID: `icon-${props.name}`},
    children: null
  }))
}));

jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(() => Promise.resolve(null)),
  setItem: jest.fn(() => Promise.resolve(null)),
  removeItem: jest.fn(() => Promise.resolve(null)),
  clear: jest.fn(() => Promise.resolve(null)),
  getAllKeys: jest.fn(() => Promise.resolve([])),
  multiGet: jest.fn(() => Promise.resolve([])),
  multiSet: jest.fn(() => Promise.resolve(null)),
  multiRemove: jest.fn(() => Promise.resolve(null)),
}));

// Mock required dependencies
jest.mock('@/services/storage', () => ({
  deleteTodo: jest.fn().mockResolvedValue(true),
  toggleTodoCompletion: jest.fn(),
  snoozeTodo: jest.fn(),
  getSnoozeDuration: jest.fn().mockResolvedValue(24),
}));

jest.mock('@/utils/toastUtils', () => ({
  showToast: jest.fn(),
}));

jest.mock('@/utils/dateUtils', () => ({
  isToday: jest.fn().mockReturnValue(false),
  isTomorrow: jest.fn().mockReturnValue(false),
  isDueSoon: jest.fn().mockReturnValue(false),
  getNextDueDate: jest.fn(),
  formatDisplayDate: jest.fn().mockReturnValue('2025-04-20'),
}));

jest.mock('@/constants/Colors', () => ({
  Colors: {
    light: {
      text: '#000',
      background: '#fff',
      cardBackground: '#fff',
      lightText: '#777',
      dueSoon: '#ff9800',
    },
    dark: {
      text: '#fff',
      background: '#000',
      cardBackground: '#222',
      lightText: '#aaa',
      dueSoon: '#ff9800',
    },
    success: 'green',
    danger: 'red',
    warning: 'orange',
    info: 'blue',
  },
}));

jest.mock('@/i18n', () => ({
  useTranslation: () => ({
    t: (key: string, params) => {
      const translations = {
        'task.edit': 'Edit Task',
        'task.deleteConfirm': 'Are you sure you want to delete this task?',
        'common.cancel': 'Cancel',
        'common.edit': 'Edit',
        'common.delete': 'Delete',
        'common.error': 'Error',
        'task.error': 'There was an error processing your task',
        'taskForm.error': 'There was an error saving your task',
        'task.hour': 'hour',
        'task.hours': 'hours',
        'task.snoozed': 'Task Snoozed',
        'task.snoozedFor': params ? `Snoozed for ${params.hours}` : 'Snoozed',
      };
      return translations[key] || key;
    },
  }),
}));

// Mock SwipeableRow component
jest.mock('../SwipeableRow', () => ({
  SwipeableRow: ({ children, onDelete, onComplete, onSnooze, isCompleted }) => (
    <div data-testid="swipeable-row" onClick={() => onDelete()}>
      {children}
      <button data-testid="delete-button" onClick={onDelete}>Delete</button>
      {onComplete && <button data-testid="complete-button" onClick={onComplete}>Complete</button>}
      {onSnooze && <button data-testid="snooze-button" onClick={onSnooze}>Snooze</button>}
    </div>
  ),
}));

// Mock Alert.alert
jest.spyOn(Alert, 'alert').mockImplementation((title, message, buttons) => {
  // Find and trigger the delete button callback if it exists
  const deleteButton = buttons?.find(button => button.text === 'Delete');
  if (deleteButton && deleteButton.onPress) {
    deleteButton.onPress();
  }
});

// Fix the react-native mock to avoid similar scope issues
jest.mock('react-native', () => {
  const rn = jest.requireActual('react-native');
  rn.Modal = jest.fn().mockImplementation(({children, visible, ...props}) => {
    if (!visible) return null;
    return {
      type: 'Modal',
      props,
      children
    };
  });
  return rn;
});

describe('TaskCard', () => {
  const mockTodo = {
    id: 'task-1',
    name: 'Test Task',
    description: 'Task description',
    completed: false,
    dueDate: '2025-04-20',
    tags: ['work', 'important'],
  };

  const mockCompletedTodo = {
    ...mockTodo,
    completed: true,
  };

  const mockEdit = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    (toggleTodoCompletion as jest.Mock).mockResolvedValue({
      success: true,
      newTodo: { ...mockTodo, completed: true },
    });
    (snoozeTodo as jest.Mock).mockResolvedValue({
      success: true,
      newDueDate: '2025-04-21',
    });
  });

  it('renders correctly (snapshot)', () => {
    const { toJSON } = render(
      <TaskCard item={mockTodo} colorScheme="light" onEdit={mockEdit} />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders task details correctly', () => {
    const { getByText } = render(
      <TaskCard item={mockTodo} colorScheme="light" onEdit={mockEdit} />
    );
    
    expect(getByText('Test Task')).toBeTruthy();
  });

  it('handles delete action', async () => {
    const { getByTestId } = render(
      <TaskCard item={mockTodo} colorScheme="light" onEdit={mockEdit} />
    );
    
    // Trigger delete
    await act(async () => {
      fireEvent.press(getByTestId('delete-button'));
    });
    
    expect(deleteTodo).toHaveBeenCalledWith('task-1');
    expect(showToast).toHaveBeenCalledWith(
      expect.stringContaining('Test Task'),
      'success'
    );
  });

  it('handles complete action', async () => {
    const { getByTestId } = render(
      <TaskCard item={mockTodo} colorScheme="light" onEdit={mockEdit} />
    );
    
    // Trigger complete
    await act(async () => {
      fireEvent.press(getByTestId('complete-button'));
    });
    
    expect(toggleTodoCompletion).toHaveBeenCalledWith('task-1');
    expect(showToast).toHaveBeenCalledWith(
      expect.stringContaining('Test Task'),
      'success'
    );
  });

  it('handles snooze action', async () => {
    const { getByTestId } = render(
      <TaskCard item={mockTodo} colorScheme="light" onEdit={mockEdit} />
    );
    
    // Trigger snooze
    await act(async () => {
      fireEvent.press(getByTestId('snooze-button'));
    });
    
    expect(getSnoozeDuration).toHaveBeenCalled();
    expect(snoozeTodo).toHaveBeenCalledWith('task-1', 24);
    expect(Alert.alert).toHaveBeenCalledWith(
      'Task Snoozed',
      expect.stringContaining('24')
    );
  });

  it('applies different styles for completed tasks', () => {
    const { container } = render(
      <TaskCard item={mockCompletedTodo} colorScheme="light" onEdit={mockEdit} />
    );
    
    // This test would check for completed task styles, but since React Native Testing Library
    // doesn't expose styles directly in the container, we'll just check the component renders
    expect(container).toBeTruthy();
  });

  it('handles long press to show edit options', () => {
    const { getByText } = render(
      <TaskCard item={mockTodo} colorScheme="light" onEdit={mockEdit} />
    );
    
    // Long press on task
    fireEvent(getByText('Test Task'), 'longPress');
    
    // Alert should be shown with edit options
    expect(Alert.alert).toHaveBeenCalledWith(
      'Edit Task',
      expect.any(String),
      expect.arrayContaining([
        expect.objectContaining({ text: 'Cancel' }),
        expect.objectContaining({ text: 'Edit' }),
        expect.objectContaining({ text: 'Delete' }),
      ])
    );
  });
});