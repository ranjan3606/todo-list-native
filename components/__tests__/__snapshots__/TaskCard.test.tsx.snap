// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskCard renders correctly (snapshot) 1`] = `
<View
  testID="swipeable-row"
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#fff",
        "borderBottomColor": "#eee",
        "borderBottomWidth": 1,
        "borderRadius": 5,
        "opacity": 1,
        "padding": 15,
      }
    }
    testID="task-card"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "marginRight": 10,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
            "paddingRight": 10,
          }
        }
        testID="completion-checkbox"
      >
        <View
          color="#ccc"
          name="circle-o"
          size={24}
          testID="icon-circle-o"
        />
      </View>
      <View
        style={
          {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 18,
                "fontWeight": "bold",
              },
              {
                "color": "#000",
              },
              false,
              false,
            ]
          }
        >
          Test Task
        </Text>
        <Text
          style={
            [
              {
                "marginTop": 5,
              },
              {
                "color": "#777",
              },
              false,
            ]
          }
        >
          Task description
        </Text>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "flexWrap": "wrap",
              "marginTop": 3,
            }
          }
        >
          <Text
            style={
              [
                {
                  "marginTop": 5,
                },
                {
                  "color": "#777",
                },
                false,
              ]
            }
          >
            task.due
          </Text>
        </View>
        <View
          style={
            {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "marginTop": 5,
            }
          }
          testID="tags-container"
        >
          <Text
            style={
              [
                {
                  "borderRadius": 8,
                  "fontSize": 11,
                  "marginBottom": 3,
                  "marginRight": 5,
                  "paddingHorizontal": 6,
                  "paddingVertical": 2,
                },
                {
                  "backgroundColor": "undefined15",
                  "color": undefined,
                  "opacity": 1,
                },
                false,
              ]
            }
          >
            work
          </Text>
          <Text
            style={
              [
                {
                  "borderRadius": 8,
                  "fontSize": 11,
                  "marginBottom": 3,
                  "marginRight": 5,
                  "paddingHorizontal": 6,
                  "paddingVertical": 2,
                },
                {
                  "backgroundColor": "undefined15",
                  "color": undefined,
                  "opacity": 1,
                },
                false,
              ]
            }
          >
            important
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="delete-button"
  >
    <Text>
      Delete
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="complete-button"
  >
    <Text>
      Complete
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="snooze-button"
  >
    <Text>
      Snooze
    </Text>
  </View>
</View>
`;
